lib/gala/payment_token.rb

```diff
@@ -37,6 +39,30 @@ def decrypt(certificate_pem, private_key_pem)
 
     class << self
 
+      def validate_signature(signature, ephemeral_public_key, data, transaction_id, application_data)
+        leaf_cert = nil
+        intermediate_cert = nil
+        p7 = OpenSSL::PKCS7.new(Base64.decode64(signature))
+        p7.certificates.each {|c|
+          c.extensions.each { |e|
+            leaf_cert = c if e.oid == '1.2.840.113635.100.6.29'
+            intermediate_cert = c if e.oid == '1.2.840.113635.100.6.2.14'
+          }
+        }
+        raise InvalidSignatureError unless leaf_cert && intermediate_cert
+        raise InvalidSignatureError unless leaf_cert.issuer == intermediate_cert.subject
+        expected_issuer = OpenSSL::X509::Name.parse "/CN=Apple Root CA - G3/OU=Apple Certification Authority/O=Apple Inc./C=US"
+        raise InvalidSignatureError unless intermediate_cert.issuer == expected_issuer
+
+        unless application_data
+          verification_string = Base64.decode64(ephemeral_public_key) + Base64.decode64(data) + [transaction_id].pack("H*")
+          # verification_string = verification_string + application_data.pack("H*") if application_data
+          store = OpenSSL::X509::Store.new
+          verified = p7.verify([], store, verification_string, OpenSSL::PKCS7::NOVERIFY )
```

Comment:
This SO post was very helpful in the process of getting this signature verified http://stackoverflow.com/questions/1999096/digital-signature-verification-with-openssl
` Checking the OpenSSL docs, we see that this unintuitive name (verify with a NOVERIFY flag?) means that you should only check the signature against the certs passed in and the certs embedded in the signature, and not attempt to verify the entire certificate chain against your trusted CA store.`
That was the final piece to fall together.  I had rewritten this line SO many times.
If you want to understand this line, the SO post would fill in all the holes.